Hardhat-ethers library linking. (#861)
nomiclabs/buidler: packages/hardhat-ethers/README.md

These helpers are added to the `ethers` object:

```typescript

> interface LibraryLinks {

Reviewer: ```suggestion
interface Libraries {
```

Reviewer: Note that these renames also need to be done in the code itself.

  [libraryName: string]: string;
}

interface FactoryOptions {
  signer?: ethers.Signer;

>   libraryLinks?: LibraryLinks;

Reviewer: ```suggestion
  libraries?: Libraries;
```

}

> function getContractFactory(name: string, signerOrFactoryOptions?: ethers.Signer | FactoryOptions): Promise<ethers.ContractFactory>;

Reviewer: Let's split this into three different signatures here, as this is documentation.

```ts
function getContractFactory(name: string): Promise<ethers.ContractFactory>;

function getContractFactory(name: string, signer: ethers.Signer): Promise<ethers.ContractFactory>;

function getContractFactory(name: string, factoryOptions?: FactoryOptions): Promise<ethers.ContractFactory>;
```

Read the documentation on the [Hardhat Runtime Environment](https://usehardhat.com/advanced/hardhat-runtime-environment.html) to learn how to access the HRE in different ways to use ethers.js from anywhere the HRE is accessible.

### Library linking

> Some contracts need to be linked with libraries before they are deployed. You can pass these library links to the `getContractFactory` function with an object like this:

Reviewer: ```suggestion
Some contracts need to be linked with libraries before they are deployed. You can pass the addresses of their libraries to the `getContractFactory` function with an object like this:
```

```js
const contractFactory = await this.env.ethers.getContractFactory(
  "Example",
  {

>     libraryLinks: {

Reviewer: ```suggestion
    libraries: {
```

      ExampleLib: "0x..."
    }
  }
);
```

This allows you to create a contract factory for the `Example` contract and link its `ExampleLib` library references to the address `"0x..."`.

> To create a contract factory, all libraries must be linked. An error will be thrown informing you of any missing library links.

Reviewer: ```suggestion
To create a contract factory, all libraries must be linked. An error will be thrown informing you of any missing library.
```
